File: two_pointers/567.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n = len(s1)
        m = len(s2)
        if n > m:
            return False
        s1 = ''.join(sorted(s1))
        # T.C : O((m-n)*logn)
        # S.C : O(1)
        for i in range(m - n + 1):
            substring = ''.join(sorted(s2[i:i+n]))
            if s1 == substring:
                return True
        return False
    

from collections import Counter


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n, m = len(s1), len(s2)
        if n > m:
            return False
        
        s1_freq = Counter(s1)
        s2_freq = Counter()
        
        i = 0
        for j in range(m):
            s2_freq[s2[j]] += 1
            if j - i + 1 > n:
                s2_freq[s2[i]] -= 1
                if s2_freq[s2[i]] == 0:
                    del s2_freq[s2[i]]
                i += 1
            if s1_freq == s2_freq:
                return True
        
        return False

File: two_pointers/567.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n = s1.length();
        int m = s2.length();
        if(n > m){
            return false;
        }
        sort(begin(s1), end(s1));
        // T.C : O((m-n)*logn)
        // S.C : O(1) 
        for(int i = 0; i <= m-n; i++){
            string substring = s2.substr(i, n);
            sort(begin(substring), end(substring));
            if(s1 == substring){
                return true;
            }
        }
        return false;
    }
};


//T.C : O(m+n)
//S.C : O(26) => O(1)
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n = s1.length();
        int m = s2.length();
        if (n > m) {
            return false;
        }
        vector<int> s1_freq(26, 0);
        vector<int> s2_freq(26, 0);
        for (char& ch : s1) { //O(n)
            s1_freq[ch - 'a']++;
        }
        int i = 0;
        int j = 0;
        while (j < m) { //O(m)
            s2_freq[s2[j] - 'a']++;
            if (j - i + 1 > n) {
                // time to shrink the window
                s2_freq[s2[i] - 'a']--;
                i++;
            }
            if (s1_freq == s2_freq) {
                return true;
            }
            j++;
        }
        return false;
    }
};